src/app/Controllers/talking.component.ts

```diff
@@ -112,23 +106,45 @@ export class TalkingComponent implements OnInit, OnDestroy {
                     }
                     t.sender.avatarURL = Values.fileAddress + t.sender.headImgFileKey;
                 });
-                if (typeof this.messages !== 'undefined' && messages.length > 0) {
+                let start = 0, end = messages.length - 1 > 0 ? messages.length - 1 : 0;
+                if (typeof this.messages !== 'undefined' && this.messages.length > 0 && messages.length > 0) {
                     if (messages[0].id === this.messages[0].id) {
                         this.noMoreMessages = true;
                     }
-                    if (!this.onBottom && messages[messages.length - 1].content !== this.messages[this.messages.length - 1].content) {
-                        this.newMessages = true;
+                    while (!getDown && start < messages.length - 1) {
+                        if (messages[start].id === this.messages[0].id) {
+                            break;
+                        }
+                        start++;
+                    }
+                    while (getDown && end >= 0) {
+                        if (messages[end].id === this.messages[this.messages.length - 1].id) {
+                            break;
+                        }
+                        end--;
+                    }
+                    if (start > 0) {
+                        this.messages.unshift(...messages.slice(0, start));
+                    }
+                    if (end < messages.length - 1) {
+                        this.messages.push(...messages.slice(end + 1));
+                        if (messages[messages.length - 1].senderId !== AppComponent.me.id) {
+                            this.newMessages = true;
+                        }
+                    } else {
+                        this.newMessages = false;
                     }
+                } else {
+                    this.messages = messages;
                 }
-                this.messages = messages;
-                if (getDown && this.onBottom) {
+                if (getDown && this.belowWindowPercent <= 0.2) {
                     setTimeout(() => {
                         this.scrollBottom(true);
                     }, 0);
                 } else if (!getDown) {
                     setTimeout(() => {
```

Comment:
I don't agree to scroll when `getDown` was false.

Reply:
That's for preventing page scroll when loading older messages.